src/documentation/programming-guide.md

```diff
@@ -532,37 +533,173 @@ tree, [2]
+windowing](#setwindowingfunction) or an [aggregation trigger](#triggers) in
+order to perform a `CoGroupByKey`. See [Unbounded PCollections](#unbounded-gbk)
+for more details.
+
+<span class="language-java">
+In the Beam SDK for Java, `CoGroupByKey` accepts a tuple of keyed
+`PCollection`s (`PCollection<KV<K, V>>`) as input. For type safety, the SDK
+requires you to pass each `PCollection` as part of a `KeyedPCollectionTuple`.
+You must declare a `TupleTag` for each input `PCollection` in the
+`KeyedPCollectionTuple` that you want to pass to `CoGroupByKey`. As output,
+`CoGroupByKey` returns a `PCollection<KV<K, CoGbkResult>>`, which groups values
+from all the input `PCollection`s by their common keys. Each key (all of type
+`K`) will have a different `CoGbkResult`, which is a map from `TupleTag<T>` to
+`Iterable<T>`. You can access a specific collection in an `CoGbkResult` object
+by using the `TupleTag` that you supplied with the initial collection.
+</span>
+<span class="language-py">
+In the Beam SDK for Python, `CoGroupByKey` accepts a dictionary of keyed
+`PCollection`s as input. As output, `CoGroupByKey` creates a single output
+`PCollection` that contains one key/value tuple for each key in the input
+`PCollection`s. Each key's value is a dictionary that maps each tag to an
+iterable of the values under they key in the corresponding `PCollection`.
+</span>
+
+The following conceptual examples use two input collections to show the mechanics of
+`CoGroupByKey`.
+
+<span class="language-java">
+The first set of data has a `TupleTag<String>` called `emailTag` and contains names
+and email addresses. The second set of data has a `TupleTag<String>` called
+`phoneTag` and contains names and phone numbers.
+</span>
+<span class="language-py">
+The first set of data contains names and email addresses. The second set of
+data contains names and phone numbers.
+</span>
 
-`CoGroupByKey` joins two or more key/value `PCollection`s that have the same key type, and then emits a collection of `KV<K, CoGbkResult>` pairs. [Design Your Pipeline]({{ site.baseurl }}/documentation/pipelines/design-your-pipeline/#multiple-sources) shows an example pipeline that uses a join. 
-
-Given the input collections below:
+```java
+// This set of data has a `TupleTag<String>` called `emailTag`.
+   "amy" -> "amy@example.com"
+   "carl" -> "carl@example.com"
+   "julia" -> "julia@example.com"
+   "carl" -> "carl@email.com"
+
+// This set of data has a `TupleTag<String>` called `phoneTag`.
+   "amy" -> "111-222-3333"
+   "james" -> "222-333-4444"
+   "amy" -> "333-444-5555"
+   "carl" -> "444-555-6666"
 ```
-// collection 1
-user1, address1
-user2, address2
-user3, address3
+```py
+{% github_sample /apache/beam/blob/master/sdks/python/apache_beam/examples/snippets/snippets_test.py tag:model_group_by_key_cogroupbykey_tuple_inputs
+%}```
 
-// collection 2
-user1, order1
-user1, order2
-user2, order3
-guest, order4
-...
-```
+<span class="language-java">
```

Comment:
I think it would be clearer if the results came after the example code.

Reply:
I can split the code snippet in two, such that the part where the tags are used to format the result are in a separate snippet. That way we can show the format of the output and then the code. I also wanted to fix the format of the output to include the tags so it's clearer to read.